src/tequila/simulators/simulator_base.py

```diff
@@ -437,8 +437,9 @@ def sample_all_z_hamiltonian(self, samples: int, hamiltonian, variables, *args,
                 Etmp += sign * count
                 n_samples += count
             E += (Etmp / samples) * paulistring.coeff
-            # small failsafe
-            assert n_samples == samples
+            # Removed this failsafe since it causes errors when simulating with samples in simulator_cirq.py for
+            # example, since the count is "filtered" and the total is sometimes less than the requested sample
+            # assert n_samples == samples
```

Comment:
looks like it actually caught unexpected behaviour. Can you give me a minimal example that reproduces the error (where n_samples != samples)

Reply:
import tequila as tq
from tequila.ml.utils_ml import preamble

initial_params_values = {'a': 0}

U = tq.gates.Rx('a', 0) + tq.gates.CNOT(1, 3)
H1 = tq.paulis.Qm(1)

objective = tq.ExpectationValue(U, H1)

cargs = {'backend': 'qulacs', 'initial_values': initial_params_values}
comped_objective, compile_args, weight_vars, w_grads, i_grads, first, second \
        = preamble(objective, cargs, None)

tq.simulate(objective=w_grads['a'][0], variables=initial_params_values, samples=100, backend='qulacs')

Reply:
Some hints I found along the way: removing tq.gates.CNOT results in no assertion throws. Also, if the objective in tq.simulate() is either objective or comped_objective, there doesn't seem to be any problems (even with CNOT).

I need tq.simulate() to work when simulating the weight gradient objective in order to obtain the numeric value of the gradient with whatever the current values of the variables are. If I should go around this some other way, please let me know.

Reply:
Since the issue is resolved: Can you revert that change here (i.e. comment the failsafe back in)